pkg/controllers/work/apply_controller.go

```diff
@@ -629,6 +672,144 @@ func buildManifestAppliedCondition(err error, action applyAction, observedGenera
+			newConfigMap, err := r.getConfigMap(ctx, obj, owner)
+			if err != nil {
+				return nil, err
+			}
+			return newConfigMap, nil
+		} else {
+			klog.ErrorS(err, "failed to retrieve config map for object", "manifest", obj.GetName(), "config map", configMapName)
+			return nil, err
+		}
+	}
+	return &configMap, nil
+}
+
+func (r *ApplyWorkReconciler) updateConfigMap(ctx context.Context, configMap *v1.ConfigMap, obj *unstructured.Unstructured) error {
+	updatedManifestHash, err := computeManifestHash(obj)
+	if err != nil {
+		// namespace for manifest can be empty
+		klog.ErrorS(err, "failed to compute new manifest hash", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+		return err
+	}
+	updatedLastModifiedConfig, err := computeModifiedConfiguration(obj)
+	if err != nil {
+		// namespace for manifest can be empty
+		klog.ErrorS(err, "failed to compute new last modified configuration", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+		return err
+	}
+	configMap.Data[manifestHashKey] = updatedManifestHash
+	configMap.Data[lastAppliedConfigKey] = updatedLastModifiedConfig
+	if err := r.spokeClient.Update(ctx, configMap); err != nil {
+		klog.ErrorS(err, "failed to update config map", "configMap", configMap.Name)
+		return err
+	}
+	return nil
+}
+
+// We can calculate the manifestHash & lastAppliedConfig again wiping out previous state or make sure the annotations in the last applied config doesn't cause issue in patch.
+func (r *ApplyWorkReconciler) migrateToConfigMap(ctx context.Context, gvr schema.GroupVersionResource, obj *unstructured.Unstructured, owner metav1.OwnerReference) error {
+	// we can also use obj.getAnnotations().
+	annots, err := metadataAccessor.Annotations(obj)
+	if err != nil {
+		return fmt.Errorf("cannot access metadata.annotations: %w", err)
+	}
+	if annots == nil {
+		return errors.New("manifest does not have any annotations")
+	}
+	// Assuming that manifest has both annotations on it
+	manifestHash, ok1 := annots[manifestHashAnnotation]
+	lastModifiedConfig, ok2 := annots[lastAppliedConfigAnnotation]
+	if ok1 || ok2 {
+		delete(annots, manifestHashAnnotation)
+		delete(annots, lastAppliedConfigAnnotation)
+		configMapName, err := r.getRandomConfigMapName(ctx)
+		if err != nil {
+			return err
+		}
+		setConfigMapNameAnnotation(configMapName, obj)
+		// update manifest
+		if _, err := r.spokeDynamicClient.Resource(gvr).Namespace(obj.GetNamespace()).Update(ctx, obj, metav1.UpdateOptions{}); err != nil {
+			klog.ErrorS(err, "failed to update manifest with config map name annotation")
+			return err
+		}
+		if err := r.createConfigMap(ctx, configMapName, owner, manifestHash, lastModifiedConfig); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ApplyWorkReconciler) getRandomConfigMapName(ctx context.Context) (string, error) {
+	generator := names.SimpleNameGenerator
+	var configMapName string
+	var configMap v1.ConfigMap
+	for {
```

Comment:
Hi Arvind! Yeah, I agree with Ealianis' comment; this is not a very good pattern, the for-loop here essentially bypasses the exponential back-off mechanism that `controll-runtime` provides, and in adverse situations the controller might get trapped in an infinite loop.

Plus, the logic here may not work as you would expect; `controller-runtime` clients use cached reads; if the cache is not fresh enough, name collisions can still occur.